core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/variables/runtime/VariableContext.scala

```diff
@@ -0,0 +1,128 @@
+/*
+ *    This file is part of ReadonlyREST.
+ *
+ *    ReadonlyREST is free software: you can redistribute it and/or modify
+ *    it under the terms of the GNU General Public License as published by
+ *    the Free Software Foundation, either version 3 of the License, or
+ *    (at your option) any later version.
+ *
+ *    ReadonlyREST is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU General Public License for more details.
+ *
+ *    You should have received a copy of the GNU General Public License
+ *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
+ */
+package tech.beshu.ror.accesscontrol.blocks.variables.runtime
+
+import cats.data.NonEmptyList
+import tech.beshu.ror.accesscontrol.blocks.rules._
+import tech.beshu.ror.accesscontrol.blocks.variables.runtime.MultiExtractable.SingleExtractableWrapper
+
+object VariableContext {
+
+  sealed trait VariableType
+
+  object VariableType {
+    trait User extends VariableType
+    trait CurrentGroup extends VariableType
+    trait Header extends VariableType
+    trait Jwt extends VariableType
+  }
+
+  trait UsingVariable[T <: Rule] {
+    def usedVariablesBy(rule: T): List[RuntimeResolvableVariable[_]]
+  }
+
+  object UsingVariable {
+    implicit val usingVariableRules: UsingVariable[Rule] = {
+      case rule: BaseSpecializedIndicesRule => rule.settings.allowedIndices.toNonEmptyList.toList
+      case rule: FilterRule => rule.settings.filter :: Nil
+      case rule: GroupsRule => rule.settings.groups.toList
+      case rule: HostsRule => rule.settings.allowedHosts.toNonEmptyList.toList
+      case rule: IndicesRule => rule.settings.allowedIndices.toNonEmptyList.toList
+      case rule: KibanaAccessRule => rule.settings.kibanaIndex :: Nil
+      case rule: KibanaIndexRule => rule.settings.kibanaIndex :: Nil
+      case rule: LocalHostsRule => rule.settings.allowedAddresses.toNonEmptyList.toList
+      case rule: UriRegexRule => rule.settings.uriPatterns.toNonEmptyList.toList
+      case rule: UsersRule => rule.settings.userIds.toNonEmptyList.toList
+      case rule: XForwardedForRule => rule.settings.allowedAddresses.toNonEmptyList.toList
+      case _ => List.empty
```

Comment:
this approach is dangerous. Do you think next developer will remember to add rule which uses variables here? I don't think so. Compiler should help us. To do so, you have to defined typeclass instance for each rule type. But you cannot just replace `_` with rest of rules, because Rule is not sealed trait.

But you can provide typeclass for each rule, eg:
```
implicit val filterRuleUsingVariable: UsingVariable[FilterRule] = rule => rule.settings.filter :: Nil
```
Then you can count on compiler which will tell you that you that there is no such typeclass instance defined for newly created Rule class.

And it's better to not defined typeclass for abstract, base classes (like BaseSpecializedIndicesRule)